<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro" >

    <xacro:include filename="inertial_macros.xacro"/>

    <material name="white">
        <color rgba="1 1 1 1" />
    </material>

    <material name="orange">
        <color rgba="1 0.3 0.1 1"/>
    </material>

    <material name="blue">
        <color rgba="0.2 0.2 1 1"/>
    </material>

    <material name="black">
        <color rgba="0 0 0 1"/>
    </material>

    <!-- BASE LINK -->

    <link name="base_link">

    </link>

    <!-- BASE_FOOTPRINT LINK -->

    <joint name="base_footprint_joint" type="fixed">
        <parent link="base_link"/>
        <child link="base_footprint"/>
        <origin xyz="0 0 0" rpy="0 0 0"/>
    </joint>

    <link name="base_footprint">
    </link>

    <!-- CHASSIS LINK -->

    <joint name="base_plate_joint" type="fixed">
        <parent link="base_link"/>
        <child link="base_plate"/>
        <origin xyz="0 0 0.25"/>
    </joint>

    <link name="base_plate">
        <visual>
            <origin xyz="0 0 0"/>
            <geometry>
                <box size="0.25 0.15 0.05"/>
            </geometry>
            <material name="white"/>
        </visual>
        <collision>
            <origin xyz="0 0 0"/>
            <geometry>
                <box size="0.25 0.15 0.05"/>
            </geometry>
        </collision>
        <xacro:inertial_box mass="2" x="0.25" y="0.15" z="0.05">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_box>
    </link>

    

    
    
    <!-- STEERING SHAFT LINK -->

    <joint name="steering_shaft_joint" type="revolute">
        <parent link="base_plate"/>
        <child link="steering_shaft"/>
        <origin xyz="0 0 -0.025" rpy="0 0 0" />
        <axis xyz="0 0 1"/>
        <dynamics damping="1.1" friction="0.5"/>
        <limit lower="${-pi/2}" upper="${pi/2}" effort="1000" velocity="${2*pi}"/>
    </joint>

    <link name="steering_shaft">
        <visual>
            <origin xyz="0 0 -0.0625"/>
            <geometry>
                <cylinder radius="0.025" length="0.125"/>
            </geometry>
            <material name="blue"/>
        </visual>
        <xacro:inertial_cylinder mass="0.5" length="0.125" radius="0.025">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>
    <!-- STEERING WHEEL LEFT LINK -->

    <joint name="steering_wheel_right_joint" type="continuous">
    <parent link="steering_shaft"/>
    <child link="steering_wheel_right"/>
    <origin xyz="-0.11 0 -0.125" rpy="0 ${pi/2} 0" />
        <axis xyz="0 0 1"/>
        <dynamics damping="0.01" friction="0.01"/>
        <!--<limit effort="1000" velocity="0.01"/>-->
        
    </joint>

    <link name="steering_wheel_right">
        <visual>
            <geometry>
                <cylinder radius="0.1" length="0.04"/>
            </geometry>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.1" length="0.04"/>
            </geometry>
        </collision>

            
        <xacro:inertial_cylinder mass="1.5" length="0.04" radius="0.1">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

     

    <joint name="steering_wheel_left_joint" type="continuous">
    <parent link="steering_shaft"/>
    <child link="steering_wheel_left"/>
    <origin xyz="0.11 0 -0.125" rpy="0 ${pi/2} 0" />
        <axis xyz="0 0 1"/>
        <dynamics damping="0.01" friction="0.01"/>
        <!--<limit effort="1000" velocity="0.01"/>-->
        
    </joint>

    <link name="steering_wheel_left">
        <visual>
            <geometry>
                <cylinder radius="0.1" length="0.04"/>
            </geometry>
            <material name="blue"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.1" length="0.04"/>
            </geometry>
        </collision>

            
            
        <xacro:inertial_cylinder mass="1.5" length="0.04" radius="0.1">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <!-- LINEAR ACTUATOR LINK -->

    <joint name="linear_actuator_joint" type="prismatic">
        <parent link="base_plate"/>
        <child link="linear_actuator"/>
        <origin xyz="0 0 0.025"/>
        <axis xyz="0 0 1"/>
        <limit lower="0.0" upper="0.12" effort="100.0" velocity="0.1"/>
        <dynamics damping="0.01" friction="0.01"/>

        
    </joint>

    <link name="linear_actuator">
        <visual>
            <origin xyz="0 0 0.05"/>
            <geometry>
                <box size="0.07 0.15 0.1"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <origin xyz="0 0 0.05"/>
            <geometry>
                <box size="0.07 0.15 0.1"/>
            </geometry>
        </collision>
        <xacro:inertial_box mass="2" x="0.07" y="0.15" z="0.1">
            <origin xyz="0 0 0.05" rpy="0 0 0"/>
        </xacro:inertial_box>
    </link>
    
    
    <!-- FRAME BOX LINK -->
    
    
    <joint name="frame_box_joint" type="fixed">
        <parent link="linear_actuator"/>
        <child link="frame_box"/>
        <origin xyz="0 0.18 -0.075" rpy="0 0 0"/>
        <dynamics damping="1.0" friction="1.0"/>

  
        
    </joint>


    <link name="frame_box">
        <visual>
            <geometry>
                <box size="0.54 0.15 0.3"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <box size="0.54 0.15 0.3"/>
            </geometry>
        </collision>
        <xacro:inertial_box mass="20" x="0.54" y="0.15" z="0.3">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_box>
    </link>
    
     <!-- LEFT FORK LINK -->
    
    
    <joint name="left_fork_joint" type="fixed">
        <parent link="frame_box"/>
        <child link="left_fork"/>
        <origin xyz="-0.19 0.08 -0.125" rpy="0 0 0"/>
  
        
    </joint>


    <link name="left_fork">
        <visual>
            <origin xyz="0 0.350 0"/>
            <geometry>
                <box size="0.16 0.7 0.05"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <origin xyz="0 0.350 0"/>
            <geometry>
                <box size="0.16 0.7 0.05"/>
            </geometry>
        </collision>
        <xacro:inertial_box mass="10" x="0.16" y="0.7" z="0.05">
            <origin xyz="0 0.350 0" rpy="0 0 0"/>
        </xacro:inertial_box>
    </link>
    
    <!-- RIGHT FORK LINK -->
    
    
    <joint name="right_fork_joint" type="fixed">
        <parent link="frame_box"/>
        <child link="right_fork"/>
        <origin xyz="0.19 0.08 -0.125" rpy="0 0 0"/>
  
        
    </joint>


    <link name="right_fork">
        <visual>
            <origin xyz="0 0.350 0"/>
            <geometry>
                <box size="0.16 0.7 0.05"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <origin xyz="0 0.350 0"/>
            <geometry>
                <box size="0.16 0.7 0.05"/>
            </geometry>
        </collision>
        <xacro:inertial_box mass="10" x="0.16" y="0.7" z="0.05">
            <origin xyz="0 0.350 0" rpy="0 0 0"/>
        </xacro:inertial_box>
    </link>
    
    
    <!-- FORK WHEEL LEFT LINK -->

    <joint name="fork_wheel_left_joint" type="continuous">
        <parent link="base_plate"/>
        <child link="fork_wheel_left"/>
        <origin xyz="-0.19 0.9 -0.2175" rpy="0 ${pi/2} 0" />
        <axis xyz="0 0 1"/>
        <dynamics damping="0.1" friction="0.1"/>
    </joint>

    <link name="fork_wheel_left">
        <visual>
            <geometry>
                <cylinder radius="0.0325" length="0.07"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.0325" length="0.07"/>
            </geometry>
        </collision>

        <xacro:inertial_cylinder mass="0.3" length="0.07" radius="0.0325">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>
    
    <!--896 mm i y till yttersta forkwheel, diameter pÃ¥ fork wheels 65 mm, bredd-->
    <!-- FORK WHEEL RIGHT LINK -->

    <joint name="fork_wheel_right_joint" type="continuous">
        <parent link="base_plate"/>
        <child link="fork_wheel_right"/>
        <origin xyz="0.19 0.9 -0.2175" rpy="0 ${pi/2} 0" />
        <axis xyz="0 0 1"/>
        <dynamics damping="0.1" friction="0.1"/>
    </joint>

    <link name="fork_wheel_right">
        <visual>
            <geometry>
                <cylinder radius="0.0325" length="0.07"/>
            </geometry>
            <material name="black"/>
        </visual>
        <collision>
            <geometry>
                <cylinder radius="0.0325" length="0.07"/>
            </geometry>
           
        </collision>

        <xacro:inertial_cylinder mass="0.3" length="0.07" radius="0.0325">
            <origin xyz="0 0 0" rpy="0 0 0"/>
        </xacro:inertial_cylinder>
    </link>

    <gazebo reference="steering_wheel_right">
        <material>Gazebo/Black</material>
        <mu1 value="0.9"/>
        <mu2 value="0.9"/>
    </gazebo>

    <gazebo reference="steering_wheel_left">
        <material>Gazebo/Black</material>
        <mu1 value="0.9"/>
        <mu2 value="0.9"/>
    </gazebo>

    <gazebo reference="fork_wheel_left">
        <material>Gazebo/Blue</material>
        <mu1 value="0.9"/>
        <mu2 value="1.0"/>
    </gazebo>
     
     <gazebo reference="fork_wheel_right">
        <material>Gazebo/Blue</material>
        <mu1 value="0.9"/>
        <mu2 value="1.0"/>
    </gazebo>
    <gazebo reference="linear_actuator_joint">
        <kp value="1000000.0"/>
        <kd value="1000000.0"/>
    </gazebo>

    <gazebo reference="frame_box">
        <material>Gazebo/Red</material>
    </gazebo>
    <!-- TRANSMISSIONS FOR ROS CONTROL -->
    
    <!-- Transmission for linear actuator -->
    <transmission name="linear_actuator_transmission">
        <type>transmission_interface/SimpleTransmission</type>
        <joint name="linear_actuator_joint">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
        </joint>
        <actuator name="linear_actuator_motor">
            <hardwareInterface>hardware_interface/EffortJointInterface</hardwareInterface>
            <mechanicalReduction>1</mechanicalReduction>
        </actuator>
    </transmission>

    <!-- STEERING SHAFT -->
    
     <!-- STEERING WHEEL-->
    
      <!-- Rear WHeels -->
     
    
    
    
</robot>
